### ISSUE TITLE
How to read function parameter values from FunctionContext for middleware

### ISSUE THREAD

**Issue by stevendarby** [Azure/azure-functions-durable-extension#3054]
### Description

If I'm using durable functions, I may have a function like this:

`public async Task RunAsync([OrchestrationTrigger] TaskOrchestrationContext taskContext, MyData myData);`

I want to write middleware that can check the value of `myData`. One way to do this currently is to first get the value of `taskContext` from the `InputBindings` metadata and then doing: `taskContext.GetInput<MyData>()`. However, there is a note on the `GetInput` method: `// NOTE: This API is redundant and may be removed in a future version.`

I believe it's described as 'redundant' because you can (optionally) take the input as a second parameter, as in the example above. However, I don't think it's possible to read that value in middleware. Is there a way?

Thanks

**Comment by geologyrocks** [comment]
As you said, I'm guessing the note is because you can now do this - (taken from [Durable functions orchestrations - Microsoft Learn](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-orchestrations?tabs=csharp-isolated#passing-multiple-parameters).)

```csharp
[Function("GetCourseRecommendations")]
public static async Task<object> RunOrchestrator(
    [OrchestrationTrigger] TaskOrchestrationContext context, 
    // note second parameter can be used to get input
    int universityYear) 
{
    string major = "ComputerScience";

    object courseRecommendations = await context.CallActivityAsync<object>(
        "CourseRecommendations",
        (major, universityYear));
    return courseRecommendations;
}
```

**Comment by fabiocav** [comment]
@stevendarby can you confirm the guidance shared by @geologyrocks answers your question? Wanted to check as your scenario might be a bit different here.

**Comment by stevendarby** [comment]
@fabiocav no, I don't think @geologyrocks intended to answer my question, and was just restating my assumption that it's marked as redundant because you can receive it as a function parameter.

My question still remains: without this method, how can you get the value in middleware? If there isn't another way, could you remove the comment and make it clear to the development team that it's still needed for middleware, so it's less unsettling to use? :D

**Comment by LockTar** [comment]
I also have a tough time to get the values of the activitytrigger or orchestrationtrigger functions. I also want to know the type. I now have the following code, but I was hoping it could be easier.

To get the type of trigger I use:
```C#
private string GetFunctionTriggerType(FunctionDefinition functionDefinition)
{
    var triggerTypeLowered = functionDefinition.InputBindings.Values.FirstOrDefault(a => a.Type.EndsWith("Trigger"))?.Type?.ToLower();

    return triggerTypeLowered;
}
```

To get a value from an ActivityTrigger or OrchestrationTrigger function I use:
```C#
private string GetMyString(ActivityType activityType, FunctionContext context, FunctionParameter functionParameter)
{
    string? myString;

    switch (activityType)
    {
        case ActivityType.Orchestration:
            context.Items.TryGetValue("__orchestrationInput__", out var command);
            myString = command?.GetType().GetProperty("MyString")?.GetValue(command) as string;
            break;
        case ActivityType.Activity:
            context.BindingContext.BindingData.TryGetValue("MyString", out var myStringObject);
            myString = myStringObject as string;
            break;
        default:
            throw new InvalidOperationException("Could not determine if function is an Orchestration or an Activity");
    }

    if (string.IsNullOrWhiteSpace(myString))
    {
        throw new InvalidOperationException($"MyString not found in the context. " +
            $"The MyString is required to run this middleware.");
    }

    return myString;
}

@fabiocav do you have an easier way???
```

**Comment by fabiocav** [comment]
Ideally, similar to the way other triggers expose APIs to work with their inputs, we'd see the same provided by Durable Functions, so I'll transfer this to the appropriate repo so the team can evaluate this and provide more information/updates.

**Comment by LockTar** [comment]
I asked my question more than a year ago and I now really have a problem. 
I'm trying to update the Azure (Durable) Function Nuget packages but this introduces a major bug inside my code.

I'm updating the packages (relevant packages only):
- `Microsoft.Azure.Functions.Worker` version `1.23.0` =>`2.51.0`
- `Microsoft.Azure.Functions.Worker.Extensions.DurableTask` version `1.1.7` => `1.16.3`

Now I can't get the input of the **orchestration** anymore in my middleware **before** `next()` is called.
As seen above, previously I could do this with:
```C#
context.Items.TryGetValue("__orchestrationInput__", out var command);
myString = command?.GetType().GetProperty("MyString")?.GetValue(command) as string;
```

I cannot get the `InstanceId` or any `parameters/command` of the orchestration.

@cgillum any idea on how to solve this? Thank you!

**Comment by stevendarby** [comment]
@LockTar my issue here may be worth checking. I’m not sure if it’s exactly the same thing but I also had trouble reading orchestration inputs in middleware after upgrading, and found a workaround, but still hoping they will support a better way in future
https://github.com/Azure/azure-functions-durable-extension/issues/3347

**Comment by LockTar** [comment]
@stevendarby  I checked your code. Were is `orchestrationContextString` (first line) coming from?

**Comment by stevendarby** [comment]
In the sample, using `string` for `T` instead of `TaskOrchestrationContext`

**Comment by LockTar** [comment]
@stevendarby Sorry don't understand what you mean but I found a different way that may be interesting. The solution is with reflection but better than nothing.

In order to get the parameters (`InputData`) **before** `next()` is called you can do the following:

Get the `OrchestratorRequest` object:
```C#
/// <summary>
/// Deserializes the base64-encoded protobuf OrchestratorRequest from the
/// IFunctionBindingsFeature InputData "context" entry.
/// Returns null if the entry is absent, malformed, or deserialization fails.
/// </summary>
private static OrchestratorRequest? GetOrchestratorRequest(FunctionContext context)
{
    try
    {
        var bindingsFeature = context.Features
            .FirstOrDefault(f => f.Key.Name == "IFunctionBindingsFeature")
            .Value;

        if (bindingsFeature is null)
        {
            return null;
        }

        var inputData = bindingsFeature.GetType()
            .GetProperty("InputData", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
            ?.GetValue(bindingsFeature) as IReadOnlyDictionary<string, object>;

        if (inputData is null || !inputData.TryGetValue("context", out var entry) || entry is null)
        {
            return null;
        }

        var base64 = entry as string
            ?? entry.GetType()
                .GetProperty("Value", BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
                ?.GetValue(entry) as string;

        if (string.IsNullOrWhiteSpace(base64))
        {
            return null;
        }

        return OrchestratorRequest.Parser.ParseFrom(Convert.FromBase64String(base64));
    }
    catch
    {
        return null;
    }
}
```

For in example to get the `InstanceId`:
```C#
/// <summary>
/// Reads the stable Durable orchestration InstanceId before next() is called.
/// The InstanceId is present on every replay via OrchestratorRequest.InstanceId.
/// </summary>
private static string? GetOrchestrationInstanceId(FunctionContext context)
{
    var request = GetOrchestratorRequest(context);
    return string.IsNullOrWhiteSpace(request?.InstanceId) ? null : request.InstanceId;
}
```

Or to get the input of the `Orchestration` as JSON:
```C#
/// <summary>
/// Reads the orchestration input JSON from the ExecutionStarted event in OrchestratorRequest.
/// On the first replay the event is in NewEvents; on subsequent replays it is in PastEvents.
/// </summary>
private static string? GetOrchestrationInputJson(FunctionContext context)
{
    var request = GetOrchestratorRequest(context);

    var executionStarted = request?.NewEvents
        .FirstOrDefault(e => e.EventTypeCase == HistoryEvent.EventTypeOneofCase.ExecutionStarted)
        ?.ExecutionStarted
        ?? request?.PastEvents
            .FirstOrDefault(e => e.EventTypeCase == HistoryEvent.EventTypeOneofCase.ExecutionStarted)
            ?.ExecutionStarted;

    var input = executionStarted?.Input;
    return string.IsNullOrWhiteSpace(input) ? null : input;
}
```

To cast the original object passed as parameter as typed object (my parameters of orchestration and sub orchestrations have the same interface):
```C#
private IFooCommand? GetOrchestrationCommand(FunctionContext context)
{
    var inputJson = GetOrchestrationInputJson(context);
    if (string.IsNullOrWhiteSpace(inputJson))
    {
        return null;
    }

    var commandParameter = GetIFooCommandFunctionParameter(context);
    return JsonSerializer.Deserialize(inputJson, commandParameter.Type) as IFooCommand;
}

private FunctionParameter GetIFooCommandFunctionParameter(FunctionContext context)
{
    // Get all the function parameters that implement the IFooCommand interface
    var parameters = context.FunctionDefinition.Parameters.Where(x => x.Type.GetInterfaces().Contains(typeof(IFooCommand)));
    if (parameters is null || !parameters.Any())
    {
        // If no parameters are found, there is something wrong with the function. 
        throw new InvalidOperationException($"No command of the type '{nameof(IFooCommand)}' found on the activity or the orchestration");
    }
    else if (parameters.Count() > 1)
    {
        // If more than one parameter is found, there is something wrong with the function. 
        throw new InvalidOperationException($"More than one command of the type '{nameof(IFooCommand)}' found on the activity or the orchestration");
    }
    else
    {
        // There is only one parameter of the type IFooCommand
        var functionParameter = parameters.Single();
        _logger.LogDebug($"Found a command parameter of the type '{nameof(IFooCommand)}' on the function. The type is '{functionParameter.Type}'");

        return functionParameter;
    }
}
```

============================================================

The result of the orchestration didn't work me neither anymore. So I solved it like this:

After the `await next(context)` call you can the result like this:
```C#
// Call the next middleware in the pipeline
await next(context);

// Get the function result — throws if the result is null.
var functionResult = context.GetInvocationResult().Value ??
    throw new InvalidOperationException("The function result is null");

// For orchestrations the invocation result is a base64-encoded OrchestratorResponse protobuf.
// Extract the actual result JSON from the CompleteOrchestrationAction embedded in that response.
// For activities the invocation result is the plain .NET return value and can be serialized directly.
var outputJson = activityType == FallbackDeploymentActivityType.Orchestration
    ? GetOrchestrationResultJson(functionResult)
    : JsonSerializer.Serialize(functionResult);
```

Search for `CompleteOrchestration`: 

```C#
/// <summary>
/// Extracts the orchestration result JSON from the base64-encoded OrchestratorResponse protobuf
/// stored in the invocation result after next() has been called.
/// The actual return value of the orchestration function is encoded inside a
/// <see cref="CompleteOrchestrationAction"/> within <see cref="OrchestratorResponse.Actions"/>.
/// Returns null when the orchestration replay did not produce a CompleteOrchestrationAction
/// (i.e. the orchestration is still awaiting activities) or when the result is empty.
/// </summary>
private static string? GetOrchestrationResultJson(object invocationResult)
{
    try
    {
        var base64 = invocationResult as string;
        if (string.IsNullOrWhiteSpace(base64))
        {
            return null;
        }

        var response = OrchestratorResponse.Parser.ParseFrom(Convert.FromBase64String(base64));
        var result = response.Actions
            .FirstOrDefault(a => a.CompleteOrchestration is not null)
            ?.CompleteOrchestration
            ?.Result;

        return string.IsNullOrWhiteSpace(result) ? null : result;
    }
    catch
    {
        return null;
    }
}
```

What do you think of it?

**Comment by stevendarby** [comment]
The parsing method looks interesting thanks, will take a look when next at computer. 

> Sorry don't understand what you mean

The first post in my linked issue shows how I used to get a TaskOrchestrationContext. Then in a reply I said I can now get a string instead, which is in base64 format. You asked how I got the string- it’s by using “string” as the type in the first post (instead of “TaskOrchestrationContext”). My “GetInputValueAsync” then returns the string in question. 

I haven’t looked too closely at your code but I think you may be over complicating some of it having missed the BindInputAsync method already provided by the framework?

**Comment by LockTar** [comment]
My integration tests are running right now. Know more tomorrow. But the code that I debugged and the result that I saw in my database was looking correct.

If you have any suggestions please let me know! 

For me the key was:
`InputData` and `OrchestratorRequest.Parser.ParseFrom(Convert.FromBase64String(base64));`

**Comment by cgillum** [comment]
A big part of the reason why writing middleware to inspect orchestrator input is so difficult is because the durable extension running in the Functions host can't simply send the input like it does with other trigger types. Rather, it must send the entire _orchestration state_, which has the input deeply embedded it.

This got more complicated in recent versions of the extension because we had to do a big refactor of the Durable middleware that runs in the worker. The change was necessary because of incompatibility with 3rd party custom middleware (like the App Configuration service middleware) that would cause silent "hangs" in orchestrations. @jviau can provide more details on that if needed.

Long-story short, we don't have a good story for using middleware to inspect orchestrator inputs, and we can't guarantee that there won't be future breaks for code that's using reflection in this way. Use reflection at your own risk!

One way this could potentially be solved is by making _durable task middleware_ a first-class concept in the .NET SDK. Then we could provide access to the input in a way that's agnostic to any Azure Functions host/worker implementation details.

/cc @torosent

**Comment by LockTar** [comment]
> Use reflection at your own risk!

Of course! I don't want to use reflection but this is the only possible way at the moment. I want to help with in example a PR but I need some guidance and direction.

> A big part of the reason why writing middleware to inspect orchestrator input is so difficult is because the durable extension running in the Functions host can't simply send the input like it does with other trigger types. Rather, it must send the entire _orchestration state_, which has the input deeply embedded it.

What is different than other trigger types that you can't simply send the input like this? And why? 

> One way this could potentially be solved is by making _durable task middleware_ a first-class concept in the .NET SDK. Then we could provide access to the input in a way that's agnostic to any Azure Functions host/worker implementation details.

What is the difference? Is this related with above question?

**Comment by jviau** [comment]
> What is different than other trigger types that you can't simply send the input like this? And why?

The payload the durable trigger sends to the worker is a serialized representation of the orchestration input, its message history, and all new events to process. The input is actually within one of those history events (`ExecutionStartedEvent`). The current durable implementation defers parsing of that input until immediately before execution. Before 'execution' was encompassed the rest of the middleware, which caused many issues due to how execution impacted the sync context. To address this execution was moved to the very end of the middleware. Unfortunately, this means it is much more difficult to access the input.

To fix it, Durable team could evaluate parsing the orchestration payload earlier in the middleware, exposing input to users, then consuming the parsed payload at the end of the pipeline. This can offer best of both worlds: access to orchestration details in middleware but not impacting middleware sync context.

@cgillum I'd leave it up to your team on evaluating the above. But happy to discuss more if you want.

**Comment by cgillum** [comment]
> To fix it, Durable team could evaluate parsing the orchestration payload earlier in the middleware, exposing input to users, then consuming the parsed payload at the end of the pipeline. This can offer best of both worlds: access to orchestration details in middleware but not impacting middleware sync context.

This is a valid option, but I feel it's a bit convoluted and requires ensuring that the user's Functions middleware executes after the Durable orchestration context parsing middleware. I was suggesting having a durable-specific middleware pipeline instead, which gives a couple benefits:

1. No need for the user to register their custom middleware in the right order
2. Durable-specific middleware can be re-used in non-Functions contexts

The downside is that it's a different type of middleware for Functions users to have to learn and it might be tricky if someone needs middleware that accesses both Functions state and Durable state. I'm not sure how to weigh the pros vs. the cons at this time.

**Comment by jviau** [comment]
> but I feel it's a bit convoluted and requires ensuring that the user's Functions middleware executes after the Durable orchestration context parsing middleware.

I understand the concern and it definitely will require the user to ensure their middleware runs after. But this is no different than something like endpoint routing in AspNetCore. They split up endpoint selection and binding and the actual invocation of the endpoint, so the user can insert their middleware between the two.

Luckily, most scenarios durable is registered via the source-gen hook, which runs as part of builder default configuration, and any manually added middleware will be after that. So, I think the vast majority of users will always have their middleware after durable.

**Comment by LockTar** [comment]
@cgillum @jviau 
So... what's next? We are now a few weeks further... Can we help with something? Should we wait for a plan?

**Comment by cgillum** [comment]
@LockTar I'm currently leaning towards a Durable Task-specific middleware feature in the SDKs (i.e. https://github.com/microsoft/durabletask-dotnet) that would meet the requirements discussed in this thread. I like the idea of solving this problem in a host-agnostic way. This is currently in the design phase. No ETA yet on when this will be available - we have to evaluate this feature work against other priorities.

**Comment by LockTar** [comment]
@cgillum thank you for the update! If you need some feedback let me know.

**Comment by cgillum** [comment]
I put together a proposal for general purpose Durable Task middleware here: https://github.com/microsoft/durabletask-dotnet/issues/721

@LockTar Here's a sample showing what it would look like for Function apps on .NET Isolated: https://github.com/Azure/azure-functions-durable-extension/tree/feature/durable-middleware-functions/samples/durable-middleware. Let me know if you think this meets your expectations.
